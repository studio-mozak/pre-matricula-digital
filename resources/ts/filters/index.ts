import { Address, GoogleMapsMarker } from '@/types';
import {
  getFormattedDate,
  getFormattedDateTime,
  getFormattedTime,
} from '@/datetime';
import {
  joinAddress,
  joinAddressLine1,
  joinAddressLine2,
  joinAddressLine3,
  markerAddress,
} from '@/util';

export interface Filters {
  formatDateTime: (date: string) => string;
  formatDate: (date: string) => string;
  formatTime: (date: string) => string;
  joinAddress: (address: Address) => string;
  joinAddressLine1: (address: Address) => string;
  joinAddressLine2: (address: Address) => string;
  joinAddressLine3: (address: Address) => string;
  markerAddress: () => GoogleMapsMarker;
}

const filters: Filters = {
  formatDateTime(date: string) {
    return getFormattedDateTime(date);
  },
  formatDate(date: string) {
    return getFormattedDate(date);
  },
  formatTime(date: string) {
    return getFormattedTime(date);
  },
  joinAddress,
  joinAddressLine1,
  joinAddressLine2,
  joinAddressLine3,
  markerAddress,
};

export default filters;
