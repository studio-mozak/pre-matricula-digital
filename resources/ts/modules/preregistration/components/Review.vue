<template>
  <div class="container-pmd">
    <h2 class="font-muli-20-primary mb-4">
      Confira seus dados antes de finalizar a inscrição:
    </h2>
    <div class="mb-3 d-sm-flex justify-content-between">
      <h3 class="font-hind-18-primary">
        Dados do(a) responsável pelo(a) aluno(a)
      </h3>
      <x-btn
        data-test="btn-edit-responsible"
        color="primary"
        outline
        class="flex-row"
        label="Editar dados"
        no-caps
        no-wrap
        @click="$emit('edit-responsible')"
      />
    </div>
    <div class="row">
      <dl v-for="field in fields.responsible" :key="field.id" class="col-6">
        <dt class="font-hind font-size-10">
          {{ field.field.name }}
        </dt>
        <dd v-if="field.type !== 'CITY'">
          {{ getField(responsible, field) }}
        </dd>
        <dd v-else>
          {{ getCity(responsible, field) }}
        </dd>
      </dl>
      <dl class="col-6">
        <dt class="font-hind font-size-10">Relação com o(a) aluno(a)</dt>
        <dd>
          {{ relationType(responsible.relationType as string) }}
        </dd>
      </dl>
      <dl class="col-sm-12">
        <dt class="font-hind font-size-10">Endereço</dt>
        <dd>
          {{ address }}
        </dd>
      </dl>
      <dl
        v-if="responsible.secondAddress && responsible.secondAddress.postalCode"
        class="col-sm-12"
      >
        <dt class="font-hind font-size-10">Endereço secundário</dt>
        <dd>
          {{ secondAddress }}
        </dd>
      </dl>
    </div>
    <div class="mt-5 mb-3 d-sm-flex justify-content-between">
      <h3 class="font-hind-18-primary">Dados do(a) aluno(a)</h3>
      <x-btn
        data-test="btn-edit-student"
        color="primary"
        outline
        class="flex-row"
        label="Editar dados"
        no-caps
        no-wrap
        @click="$emit('edit-student')"
      />
    </div>
    <div class="row">
      <dl v-for="field in fields.student" :key="field.id" class="col-6">
        <dt class="font-hind font-size-10">
          {{ field.field.name }}
        </dt>
        <dd v-if="field.type !== 'CITY'">
          {{ getField(student, field) }}
        </dd>
        <dd v-else>
          {{ getCity(student, field) }}
        </dd>
      </dl>
    </div>
    <div class="mt-5 mb-3 d-sm-flex justify-content-between">
      <h3 class="font-hind-18-primary">Dados da pré-matrícula</h3>
      <x-btn
        data-test="btn-edit-student-2"
        color="primary"
        outline
        class="flex-row"
        label="Editar dados"
        no-caps
        no-wrap
        @click="$emit('edit-student')"
      />
    </div>
    <div class="row">
      <dl class="col-6">
        <dt class="font-hind font-size-10">Série</dt>
        <dd>
          {{ grade(student.grade as string) }}
        </dd>
      </dl>
      <dl class="col-6">
        <dt class="font-hind font-size-10">Turno</dt>
        <dd>
          {{ period(student.period as string) }}
        </dd>
      </dl>
      <dl v-if="student.school" class="col-sm-12">
        <dt class="font-hind font-size-10">Escola</dt>
        <dd>
          {{ school(student.school) }}
        </dd>
      </dl>
      <dl v-if="student.secondSchool" class="col-sm-12">
        <dt class="font-hind font-size-10">Escola (lista de espera)</dt>
        <dd>
          <div>
            {{ school(student.secondSchool) }} ({{
              period(student.secondPeriod || (student.period as string))
            }})
          </div>
          <div v-for="(waitingList, index) in student.waitingList" :key="index">
            {{ school(waitingList.school) }} ({{ period(waitingList.period) }})
          </div>
        </dd>
      </dl>
    </div>
    <div class="row">
      <x-field
        v-model="accept"
        label="
        Eu autorizo o uso e tratamento dos dados pessoais acima para fins destinados a pré-matrícula na rede municipal
        "
        container-class="col-12 col-md-8 offset-md-2 mt-4"
        rules="required"
        :errors="Boolean(errors['accept'])"
        name="accept"
        type="CHECKBOX"
      />
    </div>
  </div>
</template>

<script setup lang="ts">
import {
  Fields,
  ParseFieldFromProcess,
  PreRegistrationResponsibleField,
  PreRegistrationStageProcess,
  PreRegistrationStudentField,
} from '../types';
import { computed, inject, ref } from 'vue';
import { Filters } from '@/filters';
import XBtn from '@/components/elements/buttons/XBtn.vue';
import XField from '@/components/x-form/XField.vue';
import { useGeneralStore } from '@/store/general';

defineEmits<{
  (action: 'edit-student'): void;
  (action: 'edit-responsible'): void;
}>();

const props = withDefaults(
  defineProps<{
    errors: { [key: string]: boolean };
    fields: Fields;
    responsible: PreRegistrationResponsibleField;
    student: PreRegistrationStudentField;
    process?: PreRegistrationStageProcess;
  }>(),
  {
    fields: () => ({} as Fields),
    process: () => ({} as PreRegistrationStageProcess),
    responsible: () => ({} as PreRegistrationResponsibleField),
    student: () => ({} as PreRegistrationStudentField),
  }
);

const $filters = inject('$filters') as Filters;

const store = useGeneralStore();

const accept = ref(false);

const cities = computed(() => store.getCities);

const address = computed(() => {
  return $filters.joinAddress(props.responsible.address);
});

const secondAddress = computed(() => {
  return $filters.joinAddress(props.responsible.secondAddress);
});

const relationType = (type: string) => {
  return store.relationTypes.find((r) => r.key === type)?.label;
};
const grade = (id: string) => {
  return props.process.grades.find((g) => Number(g.id) === Number(id))?.name;
};
const period = (id: string) => {
  return props.process.periods.find((p) => Number(p.id) === Number(id))?.name;
};
const school = (id: string) => {
  return props.process.schools.find((s) => Number(s.id) === Number(id))?.name;
};

const getCity = (
  payload: PreRegistrationResponsibleField | PreRegistrationStudentField,
  field: ParseFieldFromProcess
) => {
  return payload[field.key as keyof typeof payload]
    ? cities.value.find(
        (city) => city.key == payload[field.key as keyof typeof payload]
      )?.label
    : '-';
};

const getField = (
  payload: PreRegistrationResponsibleField | PreRegistrationStudentField,
  field: ParseFieldFromProcess
) => {
  return field.filter(payload[field.key as keyof typeof payload]);
};
</script>
