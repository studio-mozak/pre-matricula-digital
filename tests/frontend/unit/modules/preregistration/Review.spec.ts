import {
  Fields,
  PreRegistrationResponsibleField,
  PreRegistrationStageProcess,
  PreRegistrationStudentField,
} from '@/modules/preregistration/types';
import { VueWrapper, flushPromises, mount } from '@vue/test-utils';
import Review from '@/modules/preregistration/components/Review.vue';
import VTooltip from 'v-tooltip';
import { createTestingPinia } from '@pinia/testing';
import filters from '@/filters';
import { expect, vi } from 'vitest';

interface ReviewProps {
  errors: { [key: string]: boolean };
  fields: Fields;
  responsible: PreRegistrationResponsibleField;
  student: PreRegistrationStudentField;
  process?: PreRegistrationStageProcess;
}

const propsMock: ReviewProps = {
  errors: {},
  fields: {
    responsible: [],
    student: [],
  },
  responsible: {
    relationType: null,
    address: {
      postalCode: '',
      address: '',
      number: '',
      complement: '',
      neighborhood: '',
      lat: 0,
      lng: 0,
      city: '',
      cityIbgeCode: 0,
      stateAbbreviation: '',
    },
    useSecondAddress: false,
    secondAddress: {
      postalCode: '',
      address: '',
      number: '',
      complement: '',
      neighborhood: '',
      lat: 0,
      lng: 0,
      city: '',
      cityIbgeCode: 0,
      stateAbbreviation: '',
    },
    responsible_name: null,
  } as PreRegistrationResponsibleField,
  student: {
    grade: null,
    period: null,
    school: null,
    useSecondSchool: false,
    secondSchool: null,
    secondPeriod: null,
    match: null,
    student_name: null,
    student_date_of_birth: '2015-01-01T00:00:00.000Z',
    student_cpf: 'string',
    student_rg: 'string',
    student_marital_status: 1,
    student_place_of_birth: 1,
    student_birth_certificate: 'string',
    student_gender: 1,
    student_email: 'string',
    student_phone: 'string',
    student_mobile: 'string',
  },
  process: {
    forceSuggestedGrade: true,
    gradeAgeRangeLink: null,
    id: '1',
    messageFooter:
      'Acompanhe a publicação do resultado classificatório e convocação no site da Prefeitura.',
    name: 'PROCESSO DE TESTES 2022',
    showPriorityProtocol: true,
    grades: [
      {
        endBirth: '2016-03-31',
        id: '22',
        name: '1º Ano',
        startBirth: '2015-03-31',
      },
    ],
    periods: [
      {
        id: '1',
        name: 'Matutino',
      },
    ],
    schoolYear: { year: 2022 },
    schools: [
      {
        id: '12345',
        latitude: 0,
        longitude: 0,
        name: 'ESCOLA MUNICIPAL DE ENSINO FUNDAMENTAL',
      },
    ],
    vacancies: [
      {
        available: 8,
        grade: 6,
        period: 1,
        school: 12345,
        total: 25,
      },
    ],
    fields: [
      {
        field: {
          group: 'RESPONSIBLE',
          id: '1',
          internal: 'responsible_name',
          name: 'Nome do(a) responsável',
          options: [],
          type: 'TEXT',
        },
        id: '262',
        order: '1',
        required: true,
        weight: 0,
      },
      {
        field: {
          group: 'STUDENT',
          id: '9',
          internal: 'student_name',
          name: 'Nome',
          options: [],
          type: 'TEXT',
        },
        id: '267',
        order: '2',
        required: true,
        weight: 0,
      },
    ],
  },
};

const onMounted = async (props: ReviewProps) => {
  const wrapper = mount(Review, {
    props,
    global: {
      plugins: [createTestingPinia()],
      stubs: ['x-field'],
      directives: {
        tooltip: VTooltip,
      },
      provide: {
        $filters: filters,
      },
    },
  });

  await wrapper.vm.$nextTick();

  await flushPromises();

  return wrapper;
};

describe('Review', () => {
  let wrapper: VueWrapper<InstanceType<typeof Review>>;

  beforeEach(async () => {
    wrapper = await onMounted(propsMock);

    await wrapper.vm.$nextTick();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('the component exists', () => {
    expect(wrapper).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('emits', () => {
    test('when I click on edit responsible button, an event must be called', async () => {
      const btn = wrapper.find('[data-test="btn-edit-responsible"]');

      await btn.trigger('click');

      expect(wrapper.emitted()['edit-responsible']).toBeTruthy();
    });

    test('when I click on edit student button, an event must be called', async () => {
      const btn = wrapper.find('[data-test="btn-edit-student"]');

      await btn.trigger('click');

      expect(wrapper.emitted()['edit-student']).toBeTruthy();
    });

    test('when I click on another edit student button, an event must be called', async () => {
      const btn = wrapper.find('[data-test="btn-edit-student-2"]');

      await btn.trigger('click');

      expect(wrapper.emitted()['edit-student']).toBeTruthy();
    });
  });

  describe('grade', () => {
    test('should return the grade name', () => {
      const grade = wrapper.vm.grade('22');

      expect(grade).toBe(propsMock.process?.grades[0].name);
    });

    test('should not return the grade name', () => {
      const grade = wrapper.vm.grade('1');

      expect(grade).toBe(undefined);
    });
  });

  describe('period', () => {
    test('should return the period name', () => {
      const period = wrapper.vm.period('1');

      expect(period).toBe(propsMock.process?.periods[0].name);
    });

    test('should not return the period name', () => {
      const period = wrapper.vm.period('123');

      expect(period).toBe(undefined);
    });
  });

  describe('school', () => {
    test('should return the school name', () => {
      const school = wrapper.vm.school('12345');

      expect(school).toBe(propsMock.process?.schools[0].name);
    });

    test('should not return the school name', () => {
      const school = wrapper.vm.school('123');

      expect(school).toBe(undefined);
    });
  });

  describe('getCity', () => {
    test('should return the city name', () => {
      const city = wrapper.vm.getCity(
        propsMock.responsible,
        propsMock.fields.responsible
      );

      expect(city).toBe('-');
    });
  });
});
