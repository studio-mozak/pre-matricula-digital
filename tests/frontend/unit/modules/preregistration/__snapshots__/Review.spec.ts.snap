// Vitest Snapshot v1

exports[`Review > should match snapshot 1`] = `
<div
  class="container-pmd"
>
  <h2
    class="font-muli-20-primary mb-4"
  >
     Confira seus dados antes de finalizar a inscrição: 
  </h2>
  <div
    class="mb-3 d-sm-flex justify-content-between"
  >
    <h3
      class="font-hind-18-primary"
    >
       Dados do(a) responsável pelo(a) aluno(a) 
    </h3>
    <button
      class="x-btn x-btn-item non-selectable no-outline x-btn--outline x-btn--rectangle text-primary border-primary x-btn--actionable pmd-focusable pmd-hoverable x-btn--no-uppercase flex-row"
      data-test="btn-edit-responsible"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        class="x-btn__content text-center d-flex justify-content-center no-wrap text-no-wrap"
      >
        <span
          class="block"
        >
          Editar dados
        </span>
      </span>
      <transition-stub
        appear="false"
        css="true"
        name="x-transition--fade"
        persisted="false"
      >
        <!---->
      </transition-stub>
    </button>
  </div>
  <div
    class="row"
  >
    
    
    <dl
      class="col-6"
    >
      <dt
        class="font-hind font-size-10"
      >
        Relação com o(a) aluno(a)
      </dt>
      <dd />
    </dl>
    <dl
      class="col-sm-12"
    >
      <dt
        class="font-hind font-size-10"
      >
        Endereço
      </dt>
      <dd>
        -
      </dd>
    </dl>
    <!--v-if-->
  </div>
  <div
    class="mt-5 mb-3 d-sm-flex justify-content-between"
  >
    <h3
      class="font-hind-18-primary"
    >
      Dados do(a) aluno(a)
    </h3>
    <button
      class="x-btn x-btn-item non-selectable no-outline x-btn--outline x-btn--rectangle text-primary border-primary x-btn--actionable pmd-focusable pmd-hoverable x-btn--no-uppercase flex-row"
      data-test="btn-edit-student"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        class="x-btn__content text-center d-flex justify-content-center no-wrap text-no-wrap"
      >
        <span
          class="block"
        >
          Editar dados
        </span>
      </span>
      <transition-stub
        appear="false"
        css="true"
        name="x-transition--fade"
        persisted="false"
      >
        <!---->
      </transition-stub>
    </button>
  </div>
  <div
    class="row"
  >
    
    
  </div>
  <div
    class="mt-5 mb-3 d-sm-flex justify-content-between"
  >
    <h3
      class="font-hind-18-primary"
    >
      Dados da pré-matrícula
    </h3>
    <button
      class="x-btn x-btn-item non-selectable no-outline x-btn--outline x-btn--rectangle text-primary border-primary x-btn--actionable pmd-focusable pmd-hoverable x-btn--no-uppercase flex-row"
      data-test="btn-edit-student-2"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        class="x-btn__content text-center d-flex justify-content-center no-wrap text-no-wrap"
      >
        <span
          class="block"
        >
          Editar dados
        </span>
      </span>
      <transition-stub
        appear="false"
        css="true"
        name="x-transition--fade"
        persisted="false"
      >
        <!---->
      </transition-stub>
    </button>
  </div>
  <div
    class="row"
  >
    <dl
      class="col-6"
    >
      <dt
        class="font-hind font-size-10"
      >
        Série
      </dt>
      <dd />
    </dl>
    <dl
      class="col-6"
    >
      <dt
        class="font-hind font-size-10"
      >
        Turno
      </dt>
      <dd />
    </dl>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="row"
  >
    <x-field-stub
      allowclear="false"
      bind="[object Object]"
      containerclass="col-12 col-md-8 offset-md-2 mt-4"
      disabled="false"
      errors="false"
      label="
        Eu autorizo o uso e tratamento dos dados pessoais acima para fins destinados a pré-matrícula na rede municipal
        "
      loading="false"
      modelvalue="false"
      name="accept"
      options=""
      rules="required"
      searchable="false"
      type="CHECKBOX"
    />
  </div>
</div>
`;
